pkg/actors/internal/placement.go

```diff
@@ -0,0 +1,374 @@
+	// operationUpdateLock is the lock for three stage commit.
+	operationUpdateLock *sync.Mutex
+
+	// appHealthFn is the user app health check callback.
+	appHealthFn func() bool
+	// afterTableUpdateFn is function for post processing done after table updates,
+	// such as draining actors and resetting reminders.
+	afterTableUpdateFn func()
+
+	// shutdown is the flag when runtime is being shutdown.
+	shutdown bool
+	// shutdownConnLoop is the wait group to wait until all connection loop are done
+	shutdownConnLoop sync.WaitGroup
+}
+
+func addDNSResolverPrefix(addr []string) []string {
+	resolvers := []string{}
+	for _, a := range addr {
+		prefix := ""
+		host, _, err := net.SplitHostPort(a)
+		if err == nil && net.ParseIP(host) == nil {
+			prefix = "dns:///"
+		}
+		resolvers = append(resolvers, prefix+a)
+	}
+	return resolvers
+}
+
+// NewActorPlacement initializes ActorPlacement for the actor service.
+func NewActorPlacement(
+	serverAddr []string, clientCert *dapr_credentials.CertChain,
+	appID, runtimeHostName string, actorTypes []string,
+	appHealthFn func() bool,
+	afterTableUpdateFn func()) *ActorPlacement {
+	return &ActorPlacement{
+		actorTypes:      actorTypes,
+		appID:           appID,
+		runtimeHostName: runtimeHostName,
+		serverAddr:      addDNSResolverPrefix(serverAddr),
+		serverIndex:     0,
+
+		placementTableLock:  &sync.RWMutex{},
+		placementTables:     &hashing.ConsistentHashTables{Entries: make(map[string]*hashing.Consistent)},
+		clientCert:          clientCert,
+		operationUpdateLock: &sync.Mutex{},
+
+		appHealthFn:        appHealthFn,
+		afterTableUpdateFn: afterTableUpdateFn,
+
+		shutdown: false,
+	}
+}
+
+// Start connects placement service to register to membership and send heartbeat
+// to report the current member status periodically.
+func (p *ActorPlacement) Start() {
+	// streamConnAlive represents the status of stream channel. This must be changed in receiver loop.
+	// This flag reduces the unnecessary request retry.
+	p.streamConnAlive = true
+	p.streamConnectedCh = make(chan bool)
+	p.serverIndex = 0
+	p.shutdown = false
+	p.clientStream, p.clientConn = p.establishStreamConn()
+	if p.clientStream == nil {
+		return
+	}
+
+	// Establish receive channel to retrieve placement table update
+	p.shutdownConnLoop.Add(1)
+	go func() {
+		defer p.shutdownConnLoop.Done()
+		for !p.shutdown {
+			resp, err := p.clientStream.Recv()
+			if p.shutdown {
+				break
+			}
+
+			// TODO: we may need to handle specific errors later.
+			if !p.streamConnAlive || err != nil {
+				p.streamConnAlive = false
+				p.closeStream()
+
+				s, ok := status.FromError(err)
+				// If the current server is not leader, then it will try to the next server.
+				if ok && s.Code() == codes.FailedPrecondition {
+					p.serverIndex = (p.serverIndex + 1) % len(p.serverAddr)
+				} else {
+					log.Debugf("disconnected from placement: %v", err)
+				}
+
+				diag.DefaultMonitoring.ActorStatusReportFailed("recv", "status")
```

Comment:
Is it important that `"recv"` here matches that of line 173? If it is, please extract to constant to avoid drift.

Reply:
I would like to revisit metrics related to placement. but yes will extract it to const.